addons/binding/org.openhab.binding.yeelight/src/main/java/org/openhab/binding/yeelight/internal/discovery/DiscoveryService.java

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.yeelight.internal.discovery;
+
+import com.yeelight.sdk.device.DeviceBase;
+import com.yeelight.sdk.listeners.DeviceListener;
+import com.yeelight.sdk.services.DeviceManager;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.yeelight.YeelightBindingConstants;
+import org.openhab.binding.yeelight.internal.YeelightHandlerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DiscoveryService} is responsible for search and discovery of new devices.
+ *
+ * @author Coaster Li - Initial contribution
+ */
+public class DiscoveryService extends AbstractDiscoveryService implements DeviceListener {
+
+    private final Logger logger = LoggerFactory.getLogger(DiscoveryService.class.getSimpleName());
+
+    public DiscoveryService() {
+        super(YeelightHandlerFactory.SUPPORTED_THING_TYPES_UIDS, 2, true);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("StartBackgroundDiscovery");
+        DeviceManager.getInstance().startDiscovery();
```

Comment:
Is this starting a permanent discovery or just one-time?
Background discovery is expected to work permanently.

Reply:
For reference: https://github.com/openhab/openhab2-addons/pull/3255#discussion_r178419421

Reply:
This should run permanently. The corresponding function is here: https://github.com/coasterli/YeelightLib/blob/master/src/com/yeelight/sdk/services/DeviceManager.java#L57 although the repo is not up to date.

Reply:
> How do I change it to work permanently?

If the library does not support it, simply do not offer background discovery here. You should pass `false` as a third parameter in [this line](https://github.com/openhab/openhab2-addons/pull/3582/files#diff-58e17bcc06117206f574a0481ba60ec9R34) and remove this method here (also remove `stopBackgroundDiscovery`).

Reply:
As I wrote above it seems like the command does start a permanent discovery. The library seems to support it.

Reply:
Just had a look at the code of the lib - it is terrible as hell 🙄...

I think `startDiscovery` should better not be called as it only sends an SSDP packet out and then indefinitely waits for answers - but that does not mean that it will permanently discover devices.

It would be million times better to completely remove this discovery service implementation completely and rather register a `UPnPDiscoveryParticipant` instead. Time won't suffice to get that into 2.3, tough.

Reply:
Alright. I already suspected that after integrating the lib directly into the code base there will be a lot of more code change requests :smiley:

From my side it is ok to postpone it after 2.3, as this currently does not really meat the quality standards of openHAB I guess.

Reply:
> I think startDiscovery should better not be called as it only sends an SSDP packet out and then indefinitely waits for answers - but that does not mean that it will permanently discover devices.

Did you read this? Imho it should definitely be removed and the constructor param set to `false`.